instrumentation/okhttp/okhttp-3.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/okhttp/v3_0/OkHttp3InstrumentationModule.java

```diff
@@ -42,16 +46,69 @@ public OkHttp3InstrumentationModule() {
 
     @Override
     public Map<? extends ElementMatcher<? super MethodDescription>, String> transformers() {
-      return singletonMap(
+      Map<ElementMatcher.Junction<MethodDescription>, String> transformers = new HashMap<>();
+      transformers.put(
           isConstructor().and(takesArgument(0, named("okhttp3.OkHttpClient$Builder"))),
-          OkHttp3InstrumentationModule.class.getName() + "$OkHttp3Advice");
+          OkHttp3InstrumentationModule.class.getName() + "$OkHttp3ClientConstructorAdvice");
+      transformers.put(
+          isMethod().and(named("newBuilder").and(returns(named("okhttp3.OkHttpClient$Builder")))),
+          OkHttp3InstrumentationModule.class.getName() + "$OkHttp3ClientNewBuilderAdvice");
+      return transformers;
     }
   }
 
-  public static class OkHttp3Advice {
+  // This advice makes two guarantees:
+  // 1) The state of the builder (specifically interceptor list) is the same before and after the
+  //    OkHttpClient constructor invocation
+  // 2) The interceptor list of the created OkHttpClient has exactly one instance of the tracing
+  //    interceptor and it is in the end (assuming no other instrumentations)
+  public static class OkHttp3ClientConstructorAdvice {
     @Advice.OnMethodEnter(suppress = Throwable.class)
-    public static void addTracingInterceptor(@Advice.Argument(0) OkHttpClient.Builder builder) {
+    public static void addTracingInterceptor(
+        @Advice.Argument(0) OkHttpClient.Builder builder,
+        @Advice.Local("otelOriginalInterceptors") List<Interceptor> originalInterceptors) {
+
+      if (builder.interceptors().contains(OkHttp3Interceptors.TRACING_INTERCEPTOR)) {
+        // Potential corner case - the tracing interceptor may be in the builder due to the builder
+        // being manually constructed by adding interceptors from an existing client. In this case,
+        // save the original interceptors so we can restore them after the constructor call, and
+        // then remove all tracing interceptors before adding one as the last.
+        originalInterceptors = new ArrayList<>(builder.interceptors());
+
+        while (true) {
+          if (!builder.interceptors().remove(OkHttp3Interceptors.TRACING_INTERCEPTOR)) {
```

Comment:
```suggestion
        while (builder.interceptors().remove(OkHttp3Interceptors.TRACING_INTERCEPTOR)) {
```

Reply:
Changed. IntelliJ warned on the empty body, but I guess it is fine as it passes checkstyle.

Reply:
Changed. IntelliJ warned on empty body, but I suppose it is fine as it passes checkstyle.